src/two_sum.rs

```diff
@@ -0,0 +1,69 @@
+use std::collections::HashMap;
+/*
+ *  I did not get to choose the signature here.  I believe it should return (usize, usize)...
+ *  or even Option<(usize, usize)>.
+ */
+pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
+    /*
+     *  Per the Constraints of the problem, there will only ever be exactly one solution.
+     *  incidentally, this means that either there will only be exactly _two_ of a number, or
+     *  that the number, doubled, does not equal the target.  In either case, I only then need
+     *  a single index for a given value, as during sum finding, it either won't matter, or
+     *  whichever one I _don't_ store will just be found pointing to the one I _did store.
+     */
+    let lookup: HashMap<i32, usize> = nums.iter().copied().enumerate().map(swap).collect();
+    let (idx, first) = nums
+        .into_iter()
+        .enumerate()
+        .find(|(i, x)| {
+            lookup
+                .get(&(target - x))
+                .copied()
+                /*
+                 *  I want there to be a better pattern for this?  Option implements Eq, but
```

Comment:
I think the pattern you want is the rust special, [`map_or`](https://doc.rust-lang.org/std/option/enum.Option.html#method.map_or)!

```rust
lookup.get(&(target - x)).copied().map_or(false, |j| j != i)
```

`copied` is really cheap -- but if you felt like getting rid of it all the same, in idiomatic rust you'll also see "safe" types dereferenced to peel off a `&`. This tends to be especially true in method chains like this, where `get` returns an `Option<&T>`, but `map_or` and `map` and `filter` all _also_ take a reference and you wind up with like an `&&i32`: 

```rust
lookup.get(&(target - x)).map_or(false, |*j| j != &i) // take a reference, leave a reference
```

Since you're calling `copied` on a _single_ i32, I just... can't believe performance will be meaningfully different, more a question of personal style.

Reply:
(Just so we're on the same page, *yes*, it does make some part of me twitch that `map` returns an `Option` but `map_or` and `map_or_default` return some type `T`. I think this is because there's nothing preventing you from `T` being an `Option`, so having it "just" be generic brings considerable power. _But still_.)

Reply:
Yeah, it was more the lambda that felt weird to me - map_or is a very weird combination to me, I'd almost always want those operations separate to my eye.  Just like I prefer map/fold instead of `reduce` in JS.